qemu/tests/ksmtuned.py

```diff
@@ -0,0 +1,88 @@
+import time
+import logging
+
+from shutil import copyfile
+
+from avocado.utils import process
+
+from virttest import env_process
+from virttest.utils_test import VMStress
+from virttest.staging import utils_memory
+
+
+def run(test, params, env):
+    """
+    Check KSM can be started automaticly when ksmtuned threshold is reached
+
+    1. Get the memory of your host and the KSM_THRES_COEF
+    2. Boot a guest with memory less than KSM_THRES_COEF threshold
+    3. Get the memory used in host of process qemu-kvm
+    4. Get the free memory in host
+    5. If both the free memory size is not smaller than the threshold and guest
+        used memory + threshold is not bigger than total memory in host. Check
+        the ksm status in host. Ksm should not start in the host
+    6. Repeat step 2~5 under it broke the rule in step 5
+
+    :param test: kvm test object.
+    :param params: Dictionary with test parameters.
+    :param env: Dictionary with the test environment.
+    """
+    def check_ksm(mem, stress=False):
+        """
+        :param mem: Boot guest with given memory
+        :param stress: Load stress or not
+        """
+        params['mem'] = mem // 1024
+        params['start_vm'] = 'yes'
+        vm_name = params['main_vm']
+        env_process.preprocess_vm(test, params, env, vm_name)
+        vm = env.get_vm(vm_name)
+        vm.verify_alive()
+        if stress:
+            params['stress_args'] = ('--cpu 4 --io 4 --vm 2 --vm-bytes %sM' %
+                                     (int(params['mem']) // 2))
+            stress_test = VMStress(vm, "stress", params)
+            stress_test.load_stress_tool()
+            time.sleep(10)
+        qemu_pid = vm.process.get_pid()
+        qemu_used_mem = int(process.getoutput(params['cmd_get_qemu_used_mem'] %
+                                              qemu_pid, shell=True)) * 4 * 1024
+        free_mem_host = utils_memory.freememtotal()
+        vm.destroy()
+        ksm_status = process.getoutput(params['cmd_check_ksm_status'])
+        logging.info('The ksm threshold is %s, the memory allocated by qemu is'
+                     ' %s, and the total free memory on host is %s.' %
+                     (ksm_thres, qemu_used_mem, free_mem_host))
+        if (free_mem_host >= ksm_thres and
+                (qemu_used_mem + ksm_thres) <= total_mem_host):
+            if ksm_status != '0':
+                test.fail('Ksm should not start.')
+            if stress:
+                test.error('The host resource is not consumed as expected.')
+        elif ksm_status == '0':
+            test.fail('Ksm should start but it does not.')
+
+    total_mem_host = utils_memory.memtotal()
+    free_mem_host = utils_memory.freememtotal()
+    ksm_thres = process.getoutput(params['cmd_get_thres'], shell=True)
+    ksm_thres = int(total_mem_host *
+                    (int(re.findall('\\d+', ksm_thres)[0]) / 100))
+    guest_mem = (free_mem_host - ksm_thres) // 2
+    check_ksm(guest_mem)
```

Comment:
Suggest to check ksmtuned service status first, start it if it's disabled, and restore it in finally.